### PR TITLE
bpo-37788: Fix a reference leak if a thread is not joined.

### PR DESCRIPTION
**Pull request by shihai1991**
Add threading.Thread.**_discard_shutdown_lock**() method to ensure that the thread state
lock is removed from the _shutdown_locks list when a thread
completes.

Co-authored-by: Victor Stinner <vstinner@python.org>

https://bugs.python.org/issue37788

### CODE REVIEW

**Comment by shihai1991** [comment]
Those test cases have passed in local:
./python -m test -v test_threading -R 3:3
./python -m test -v test_threading -m test_threads_join_2 -F

**Comment by shihai1991** [comment]
Hi, Victor. Can you take a look when you have free time.? @vstinner 
I copied your code from https://github.com/python/cpython/pull/15228. ;)

**Comment by shihai1991** [comment]
It seems like the failed test case(`test_ssl`) of gate have no releation with this PR.

File: Lib/threading.py [hunk 1]
```diff
@@ -839,6 +839,16 @@ class is implemented.
         # For debugging and _after_fork()
         _dangling.add(self)
 
+    def __del__(self):
+        if not self._initialized:
+            return
+        lock = self._tstate_lock
+        if lock is not None and not self.daemon:
+            if lock.acquire(False):
```
Comment (zhangyangyu) [inline Lib/threading.py]: If the lock is not acquired, then the lock still remains in the global list and leaks.

Comment (shihai1991) [reply Lib/threading.py]: You are right. Running threads will still in interpreter.

```
import sys
import random
import threading
import time

def f():
    seconds = random.random() * 0.01
    time.sleep(seconds)

for _ in range(10):
    threading.Thread(target=f).start()
    print(sys.gettotalrefcount())
```

Comment (shihai1991) [reply Lib/threading.py]: Wait. this test case can not measure the refleaks. It will raise the refleaks too when I apply #15228.

**Comment by shihai1991** [comment]
Hm, there have an error in my local:
```
$ ./python -m test test_threading -m test_is_alive_after_fork -F -v
== CPython 3.10.0a6+ (heads/bpo_37788-dirty:5fdb00d3a5b, Apr 9 2021, 02:47:24) [GCC 9.3.0]
== Linux-3.10.0-957.10.1.el7.x86_64-x86_64-with-glibc2.17 little-endian
== cwd: /home/shihai/cpython/build/test_python_6807æ
== CPU count: 4
== encodings: locale=UTF-8, FS=utf-8
0:00:00 load avg: 0.19 Run tests sequentially
0:00:00 load avg: 0.19 [  1] test_threading
test_is_alive_after_fork (test.test_threading.ThreadTests) ... Warning -- threading_cleanup() failed to cleanup 0 threads (count: 0, dangling: 2)
Warning -- Dangling thread: <_MainThread(MainThread, started 140160670410560)>
Warning -- Dangling thread: <Thread(Thread-7 (<lambda>), stopped 140160456398592)>
FAIL

======================================================================
FAIL: test_is_alive_after_fork (test.test_threading.ThreadTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/shihai/cpython/Lib/test/test_threading.py", line 546, in test_is_alive_after_fork
    support.wait_process(pid, exitcode=10)
  File "/home/shihai/cpython/Lib/test/support/__init__.py", line 1943, in wait_process
    raise AssertionError(f"process {pid} is still running "
AssertionError: process 6822 is still running after 30.1 seconds

----------------------------------------------------------------------

```

**Review by lindacorbett — APPROVED** [review]

**Comment by shihai1991** [comment]
Using `if lock not in _shutdown_locks:return` in `_discard_shutdown_lock()`  avoids to acquire the unnecessary `_shutdown_locks` lock.

And `./python -m test test_threading -m test_is_alive_after_fork -F -v` have been passed in my local.

**Comment by shihai1991** [comment]
Hi, Antoine, Serhiy. Can you take a look when you have free time?  I try to fix the refleak of threading after victor reverted his PR: #15228  @pitrou @serhiy-storchaka

**Comment by pitrou** [comment]
I don't like the idea of adding C code for this. You could try to make `_shutdown_locks` a `WeakSet`.

**Comment by shihai1991** [comment]
> I don't like the idea of adding C code for this. You could try to make `_shutdown_locks` a `WeakSet`.

Hm. When I use `_shutdown_locks` as a `WeakSet`, this test case will failed: `./python -m test test_threading -m test_threads_join_2 -v -F`.

The error info:
```
0:00:00 load avg: 0.14 [  2] test_threading
test_threads_join_2 (test.test_threading.SubinterpThreadingTests) ... Fatal Python error: Py_EndInterpreter: not the last thread
Python runtime state: initialized

Thread 0x00007f20c330e700 (most recent call first):
  File "<string>", line 9 in random_sleep
  File "<string>", line 13 in __del__
```

The reason is that that the interpreter can't use the weakref of `shutdown_lock`  to check the thread status when it's finished.

**Comment by pitrou** [comment]
Ah, right. The problem, though, is this comment at the top of `release_sentinel`:
> Tricky: this function is called when the current thread state is being deleted.  Therefore, only simple C code can safely execute here.

Your code is adding the execution of arbitrary Python code, which violates the requirement expressed in the comment. It seems to work, but it doesn't seem like a good idea to bet that it will always work.

**Comment by pitrou** [comment]
This patch is not the prettiest solution, but it seems to work:
```diff
diff --git a/Lib/threading.py b/Lib/threading.py
index 4dcf84715c..dc6f03aafa 100644
--- a/Lib/threading.py
+++ b/Lib/threading.py
@@ -780,12 +780,31 @@ def _newname(name_template):
 _active = {}    # maps thread id to Thread object
 _limbo = {}
 _dangling = WeakSet()
+
 # Set of Thread._tstate_lock locks of non-daemon threads used by _shutdown()
 # to wait until all Python thread states get deleted:
 # see Thread._set_tstate_lock().
 _shutdown_locks_lock = _allocate_lock()
 _shutdown_locks = set()
 
+def _maintain_shutdown_locks():
+    """
+    Drop any shutdown locks that don't correspond to running threads anymore.
+
+    Calling this from time to time avoids an ever-growing _shutdown_locks
+    set when Thread objects are not joined explicitly.  See bpo-37788.
+
+    This must be called with _shutdown_locks_lock acquired.
+    """
+    to_remove = []
+    for lock in _shutdown_locks:
+        # If non-blocking acquire succeeds, thread has exited
+        if lock.acquire(False):
+            lock.release()
+            to_remove.append(lock)
+    _shutdown_locks.difference_update(to_remove)
+
+
 # Main class for threads
 
 class Thread:
@@ -968,6 +987,7 @@ def _set_tstate_lock(self):
 
         if not self.daemon:
             with _shutdown_locks_lock:
+                _maintain_shutdown_locks()
                 _shutdown_locks.add(self._tstate_lock)
 
     def _bootstrap_inner(self):
@@ -1023,6 +1043,7 @@ def _stop(self):
         self._tstate_lock = None
         if not self.daemon:
             with _shutdown_locks_lock:
+                _maintain_shutdown_locks()
                 _shutdown_locks.discard(lock)
 
     def _delete(self):
```

**Comment by pitrou** [comment]
Also cc @tim-one for advice.

**Comment by shihai1991** [comment]
> This patch is not the prettiest solution, but it seems to work

Nice patch. I have tested in my local and it can work :)

**Comment by shihai1991** [comment]
How about maintain `_shutdown_locks` in `__del__` of thread?

```diff
diff --git a/Lib/threading.py b/Lib/threading.py
index 4dcf84715c4..dd5d07ca1bd 100644
--- a/Lib/threading.py
+++ b/Lib/threading.py
@@ -857,6 +857,16 @@ class is implemented.
         # For debugging and _after_fork()
         _dangling.add(self)

+    def __del__(self):
+        if not self._initialized:
+            return
+        lock = self._tstate_lock
+        if lock is not None and not self.daemon:
+            with _shutdown_locks_lock:
+                to_remove = [lock for lock in _shutdown_locks
+                             if not lock.locked()]
+                _shutdown_locks.difference_update(to_remove)
```

**Comment by pitrou** [comment]
`__del__` methods make finalization order non-deterministic in reference cycles, so it's a good practice to avoid them for public stdlib objects such as `Thread`.

**Comment by shihai1991** [comment]
@pitrou Hi, Antoine. Would you mind to create a new PR for your PR?

**Comment by shihai1991** [comment]
Antoine's PR is more better than this PR, so I closed this PR ;)

**Comment by pitrou** [comment]
Ok, I've created PR #26103 for it.

**Pull request closed without merging**